handle postgres GSSEncRequest startup message

GavinFrazar opened the pull request:
Closes #6284

## What
1. Recognize the "GSSEncRequest" message in multiplexer as postgres protocol.
2. Change postgres engine to handle this request at connection startup and respond 'N' for "not supported" to make client choose SSL encryption instead or disconnect.
3. Add `gssencmode=disable` to the ~/.pg_service.conf file we save on db login.

## Why
We don't support GSS encryption, but this message is sent at startup and we can easily respond with 'N' to let the client try SSL. This way they don't just automatically get an error.

GavinFrazar committed:
Recognize GSSEncRequest as postgres proto

GavinFrazar committed:
Respond to GSSEncRequest with 'N' for 'not supported' in postgres engine

GavinFrazar committed:
Add test for GSSEncRequest to postgres via proxy

GavinFrazar committed:
Save gssencmode=disable in .pg_service.conf and test it

smallinsky commented on lib/srv/db/proxy_test.go:
```diff
@@ -54,6 +57,31 @@ func TestProxyProtocolPostgres(t *testing.T) {
+
+			// Check that we can send a GSSEncRequest and get an 'N' response indicated server does not accept
+			// GSS encryption.
+			conn, err := net.Dial("tcp", proxy.Address())
+			require.NoError(t, err)
+			defer func() {
+				require.NoError(t, conn.Close())
+			}()
+			req := pgproto3.GSSEncRequest{}
+			msg := req.Encode(nil)
+			bytesWritten, err := conn.Write(msg)
+			require.NoError(t, err)
+			require.Equal(t, len(msg), bytesWritten)
+			buf := make([]byte, 512)
+			bytesRead, err := conn.Read(buf)
+			require.NoError(t, err)
+			require.Equal(t, []byte("N"), buf[:bytesRead], "server should respond with 'N' to indicate GSS encryption is not supported'")
+			sslReq := pgproto3.SSLRequest{}
+			msg = sslReq.Encode(nil)
+			bytesWritten, err = conn.Write(msg)
+			require.NoError(t, err, "server should not have closed conn after rejecting GSS")
+			require.Equal(t, len(msg), bytesWritten)
+			bytesRead, err = conn.Read(buf)
+			require.NoError(t, err)
+			require.Equal(t, []byte("S"), buf[:bytesRead], "server should respond with 'S' to indicate SSL encryption is supported'")
```
nit: can we extract this GSS check to a separate function ?

Also from the test desc: 
> TestProxyProtocolPostgres ensures that clients can successfully connect to a
// Postgres database when Teleport is running behind a proxy that sends a proxy
// line.

Can we consider adding a new test instead of extending current test ?

GavinFrazar replied:
Will address this after fixing what we suggested in the other repo.

GavinFrazar replied:
Refactored into a (substantial) test that covers all code paths in `handleStartup`. I also made the tests run in parallel which sped things up substantially for tests running `lib/srv/db/proxy_test.go` from 15s runtime to 4.5 seconds locally. All tests still passed and I ran with the race detector on.

GavinFrazar committed:
Check for repeated startup encryption request

GavinFrazar committed:
Refactor postgres startup test code into its own test

* Test covers all code paths in the proxy startup handler

GavinFrazar committed:
Make postgres proxy tests run in parallel

* Reduced test execution time from 15 seconds to 4.5 seconds.
* Ran with race detector and all tests passed.

GavinFrazar commented:
@smallinsky @hugoShaka I added substantial changes to the test code in `lib/srv/db/proxy_test.go` and I refactored `lib/srv/db/postgres/proxy.go:handleStartup` to use a loop instead of recursion. Please take another look and approve. I don't want to merge this without letting you guys review these additional changes.

GavinFrazar committed:
Make sure we read at least one byte

GavinFrazar committed:
Fix lint

GavinFrazar committed:
Fix lint

GavinFrazar committed:
Change err to bad parameter

GavinFrazar committed:
Update test

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/handle_unexpected_postgres_startup_message

GavinFrazar commented:
@smallinsky you added a backport label for v8, but do we need to do that since we only support the last 3 versions (v9, v10, v11)?
